packages/firefly/src/AppRouter.tsx

```diff
@@ -0,0 +1,188 @@
+                logger.debug(`[shell] Modules are ${areModulesReady ? "ready" : "registered"}, waiting for MSW to start...`);
+            }
+
+            if (!areModulesRegistered && !areModulesReady && isMswStarted) {
+                logger.debug("[shell] MSW is started, waiting for the modules...");
+            }
+        }
+    }, [logger, areModulesRegistered, areModulesReady, isMswStarted, waitForMsw]);
+
+    useEffect(() => {
+        // Don't go further if no handler has been provided to load public data.
+        if (onLoadPublicData) {
+            if ((areModulesRegistered || areModulesReady) && isMswStarted) {
+                if (!isPublicDataLoaded) {
+                    logger.debug("[shell] Loading public data.");
+
+                    const abordController = new AbortController();
+
+                    onLoadPublicData(abordController.signal).finally(() => {
+                        setIsPublicDataLoaded(true);
+
+                        logger.debug("[shell] Public data has been loaded.");
+                    });
+
+                    return () => {
+                        abordController.abort();
+                    };
+                }
+            }
+        }
+    }, [logger, areModulesRegistered, areModulesReady, isMswStarted, isPublicDataLoaded, onLoadPublicData]);
+
+    const isActiveRouteProtected = useIsRouteMatchProtected(location);
+
+    useEffect(() => {
+        // Don't go further if no handler has been provided to load protected data.
+        if (onLoadProtectedData) {
+            if ((areModulesRegistered || areModulesReady) && isMswStarted) {
+                if (isActiveRouteProtected) {
+                    if (!isProtectedDataLoaded) {
+                        logger.debug(`[shell] Loading protected data as "${location.pathname}" is a protected route.`);
+
+                        const abordController = new AbortController();
+
+                        onLoadProtectedData(abordController.signal).finally(() => {
+                            setIsProtectedDataLoaded(true);
+
+                            logger.debug("[shell] Protected data has been loaded.");
+                        });
+
+                        return () => {
+                            abordController.abort();
+                        };
+                    }
+                } else {
+                    logger.debug(`[shell] Not loading protected data as "${location.pathname}" is a public route.`);
+                }
+            }
+        }
+    }, [logger, location, areModulesRegistered, areModulesReady, isMswStarted, isActiveRouteProtected, isProtectedDataLoaded, onLoadProtectedData]);
+
+    useEffect(() => {
+        // Don't go further if no handler has been provided to complete the registration.
+        if (onCompleteRegistration) {
+            if (areModulesRegistered && isMswStarted && isPublicDataLoaded) {
+                if (!areModulesReady) {
+                    onCompleteRegistration();
```

Comment:
Don't you want to do the same .finally(() => {
                            logger.debug("[shell] Registration has been completed.");
}

Reply:
The `completeRegistrations` function is already logging something similar.